libkineto/src/ActivityProfiler.cpp

```diff
@@ -114,7 +117,11 @@ bool ActivityProfiler::applyNetFilterInternal(const std::string& name) {
 }
 
 // This has dependence on CuptiActivityInterface
+#ifdef HAS_ROCTRACER
+ActivityProfiler::ActivityProfiler(RoctracerActivityInterface& cupti, bool cpuOnly)
```

Comment:
Can we instead create a pure virtual class IDeviceActivityProfiler, with generic function names like enableActivities() and so on? This will be resolved with my ongoing refactoring, but would be nice to at least avoid additional #ifdefs like these.

Reply:
Yeah, it's pretty clear that this stage of development wasn't really done with multiple backends in mind.  I was hoping the "refactor" would make this fit together more cleanly.  I didn't want to hijack the refactor... so there is some not-so-elegant stuff in there.  IMO there shouldn't be an '#ifdef Anything' in any of the files save once per header.  That should be handled in the build.  Cleaning this up will be an effort.  As you know, there is cupti code and cupti cruft everywhere.  E.g. ActivityProfiler (used by cpu, cupti, and roctracer) has cupti code all through it.  Ideally that would be ifdef-clean and cupti/roctracer concerns would be layered on with separate source files.

We could create this proposed interface now, but it means very little until we get the cupti concerns out of ActivityProfiler.  There will still be as many ifdefs (you could change some to dynamic_cast<> checks) :)  Keep in mind I didn't add any "ifdef cupti" to that file, those are all there from the original separation of concerns problem.  So those can't go away until a well thought-out refactor takes place.   As a simple example, to generalize enableActivity() I will have to decide on and define the actual Activities enum. Then go change all the existing code that uses the cupti provided enums.

So if I do that now then undeniably the refactor is creeping into a roctracer support PR.  Honestly, I was seeing this "not so elegant" integration as a tool for you to direct your refactoring.  Having the integration in place should produce a better refactor as the issues are real instead of hypothetical.

So the full refactor should not be in this PR, clearly.  We can pull this small and incomplete part of it in right now, but I'd personally rather not.  Your call.

Reply:
Fair. I was under the impression that CUPTI and the ROC tracer APIs were very similar, hence the comments. But I want to avoid more work for no gain obviously. What's your timeline for this, do you have a special reason to get this in for 1.10 or is it just as soon as possible?

Reply:
Yes, the cupti and roctracer interfaces are superficially very similar.  It's in the details that implementations diverge.  And of course all the types, constants, etc. are going to have to be generalized.  That generalization looks silly when you have a single backend (hence you skipped it) but required when you have two or more.

Oh, indeed ActivityType is already generalized (contrary to what I said in the previous comment).   Good.  I got confused with the ActivityController bits that are handling cupti-specific stuff.

We would like to get this into 1.10 because we have some big installations using it.  We've been maintaining a 1.09 fork with this enabled.  Would really like to avoid that support drag and make it easier for the users to get.  And yeah, I'll would have to create the 1.10 fork personally. :P